src/heuristics/WordNet.py

```diff
@@ -0,0 +1,62 @@
+import logging
+
+from nltk.corpus import wordnet
+import nltk
+import itertools
+
+from heuristics.Heuristics import AbstractHeuristic
+
+log = logging.getLogger(__name__)
+
+# Source: https://stackoverflow.com/a/20037408/1556343
+def flatmap(func, *iterable):
+    return itertools.chain.from_iterable(map(func, *iterable))
+
+class WordNetHeuristic(AbstractHeuristic):
+    def __init__(self):
+        self.wordNet = None
+        self.api = None
+        self.goal = None
+        self.goal_synsets = None
+
+    def get_first_synset(self, word):
+        word_synsets = wordnet.synsets(word)
+        if not word_synsets:
+            return []
+        return [word_synsets[0]]
+
+    def get_synsets(self, word):
+        """
+        Get synsets for a word.
+        If the word has a WordNet node, return it's own synsets.
+        If the word doesn't, return all the synsets of it's neighbors.
+        """
+        word_synsets = wordnet.synsets(word)
+        if word_synsets:
+            return [word_synsets[0]]
+        try:
+            neighbors = self.api.get_text_and_links(word)[1]
+            return list(flatmap(lambda title: self.get_first_synset(title), neighbors))
+        except OSError:
+            return []
+
+    def setup(self, api, start, goal):
+        log.info("Initializing WordNet heuristic")
+        # Ensure that we have the WordNet corpus
+        # This will download the corpus on the first run, but not on subsequent runs
+        nltk.download('wordnet')
```

Comment:
Is there a way to log when we're downloading the corpus vs loading an existing one? That way it's less surprising if it hangs for a while.

Reply:
It only hangs for a second or two when this happens. It's a small file. I don't think we need to worry about this.